diff --git a/firmware/fw_base.ldS b/firmware/fw_base.ldS
index 0ac75f2..f2433b3 100644
--- a/firmware/fw_base.ldS
+++ b/firmware/fw_base.ldS
@@ -73,9 +73,74 @@
 		. = ALIGN(8);
 		PROVIDE(_bss_end = .);
 	}
+    .embed : {
+        /* 4 bytes magic number !emb */
+    	BYTE(33)
+    	BYTE(101)
+    	BYTE(109)
+    	BYTE(98)
+    	PROVIDE(_protocol_version = .);
+    	. += 0x1;
 
+    	PROVIDE(_rt_root_public_key = .);
+    	. += 0x20;
+
+    	PROVIDE(_eapp_root_public_key = .);
+    	. += 0x20;
+
+		PROVIDE(_rt_root_enc_key = .);
+    	. += 0x20;
+
+    	PROVIDE(_eapp_rt_enc_key = .);
+    	. += 0x20;
+  	}
+
+	.runtime :
+	{
+		PROVIDE(_runtime_start = .);
+		*(.runtime)
+		. = ALIGN(8);
+		PROVIDE(_runtime_end = .);
+	}
 	/* End of the read-write data sections */
 
 	. = ALIGN(0x1000); /* Need this to create proper sections */
 
 	PROVIDE(_fw_end = .);
+
+  * # Sanctum params */
+  /* ================ */
+  . = 0x801ff000; /* the last page before the payload */
+
+  /* ## manufacturer_keys : */
+
+  /* 32 Bytes : manufacturer public key */
+  PROVIDE( sanctum_m_public_key = . );
+  . += 0x20;
+
+  /* 32 Bytes : device public key */
+  PROVIDE( sanctum_dev_public_key = . );
+  . += 0x20;
+
+  /* 64 Bytes : device secret key */
+  PROVIDE( sanctum_dev_secret_key = . );
+  . += 0x40;
+
+  /* ## security_monitor_keys : */
+
+  /* 64 Bytes : security monitor hash */
+  PROVIDE( sanctum_sm_hash = . );
+  . += 0x40;
+
+  /* 32 Bytes : security monitor public key */
+  PROVIDE( sanctum_sm_public_key = . );
+  . += 0x20;
+
+  /* 64 Bytes : security monitor secret key */
+  PROVIDE( sanctum_sm_secret_key = . );
+  . += 0x40;
+
+  /* 64 Bytes : security monitor's signature by device */
+  PROVIDE( sanctum_sm_signature = . );
+  . += 0x40;
+
diff --git a/firmware/fw_payload.S b/firmware/fw_payload.S
index 1ef121e..b26016b 100644
--- a/firmware/fw_payload.S
+++ b/firmware/fw_payload.S
@@ -95,3 +95,11 @@ payload_bin:
 #else
 	.incbin	FW_PAYLOAD_PATH
 #endif
+
+	.section .runtime, "ax", %progbits
+	.align 4
+	.globl runtime
+runtime:
+#ifdef RUNTIME_PATH
+	.incbin	RUNTIME_PATH
+#endif
diff --git a/firmware/objects.mk b/firmware/objects.mk
index b2ace75..4dc7c70 100644
--- a/firmware/objects.mk
+++ b/firmware/objects.mk
@@ -41,6 +41,11 @@ else
 FW_PAYLOAD_PATH_FINAL=$(platform_build_dir)/firmware/payloads/test.bin
 endif
 firmware-genflags-$(FW_PAYLOAD) += -DFW_PAYLOAD_PATH=\"$(FW_PAYLOAD_PATH_FINAL)\"
+
+ifdef RUNTIME_PATH
+firmware-genflags-$(FW_PAYLOAD) += -DRUNTIME_PATH=\"$(RUNTIME_PATH)\"
+endif
+
 ifdef FW_PAYLOAD_OFFSET
 firmware-genflags-$(FW_PAYLOAD) += -DFW_PAYLOAD_OFFSET=$(FW_PAYLOAD_OFFSET)
 endif
